webapp/src/idbworkspace.ts

```diff
@@ -267,12 +288,36 @@ async function getDbAsync(prefix = "__default") {
     }
 }
 
-async function copyTableEntriesAsync(fromDb: pxt.BrowserUtils.IDBWrapper, toDb: pxt.BrowserUtils.IDBWrapper, storeName: string, dontOverwrite: boolean) {
+async function copyTableEntriesAsync(fromDb: pxt.BrowserUtils.IDBWrapper, toDb: pxt.BrowserUtils.IDBWrapper, storeName: string, dontOverwrite: boolean, fromPrefix: string, toPrefix: string) {
+    const migrationDb = await getMigrationDbAsync();
+
     for (const entry of await fromDb.getAllAsync<any>(storeName)) {
+        const key = migrationDbPrefixUpgradeKey(fromPrefix, toPrefix, entry.id);
+        if (await migrationDb.getAsync(storeName, key)) continue;
+
         const existing = dontOverwrite && !!(await toDb.getAsync(storeName, entry.id));
 
         if (!existing) {
             await toDb.setAsync(storeName, entry);
+            await migrationDb.setAsync(storeName, {
+                id: key
+            });
+        }
+    }
+}
+
+async function populatePrefixMigrationDb(fromDb: pxt.BrowserUtils.IDBWrapper, toDb: pxt.BrowserUtils.IDBWrapper, storeName: string, fromPrefix: string, toPrefix: string) {
+    const migrationDb = await getMigrationDbAsync();
+
+    for (const entry of await fromDb.getAllAsync<any>(storeName)) {
+        const key = migrationDbPrefixUpgradeKey(fromPrefix, toPrefix, entry.id);
+        if (await migrationDb.getAsync(storeName, key)) continue;
+
+        // If this header id was actually migrated, add an entry for it in the migration db
+        if (await toDb.getAsync(storeName, entry.id)) {
```

Comment:
Could we combine this and the check above and use short circuiting to avoid this check if we get something from the `migrationDb`? Or do we want to keep it separate in case we want to do something else if the `migrationDb` exists?

Reply:
this check is already skipped if there is an entry from the migrationdb thanks to the continue